migration/sql-files/080-remove-unknown-link-types.sql

```diff
@@ -0,0 +1,95 @@
+-- unknown link types.
+SET LOCAL linktypes.unknown_bug_blocker = 'aad2a4ad-d601-4104-9804-2c977ca2e0c1';
+SET LOCAL linktypes.unknown_related = '355b647b-adc5-46b3-b297-cc54bc0554e6';
+SET LOCAL linktypes.unknown_parenting = '7479a9b9-8607-46fa-9535-d448fa8768ab';
+
+-- Re-assign any existing link to correct type but avoid those links that
+-- already exist.
+UPDATE work_item_links l 
+    SET link_type_id = current_setting('linktypes.bug_blocker')::uuid 
+    WHERE link_type_id = current_setting('linktypes.unknown_bug_blocker')::uuid
+    AND NOT EXISTS(
+        SELECT * FROM work_item_links
+        WHERE
+            link_type_id = current_setting('linktypes.bug_blocker')::uuid
+            AND source_id = l.source_id
+            AND target_id = l.target_id
+        );
+UPDATE work_item_links l 
+    SET link_type_id = current_setting('linktypes.related')::uuid 
+    WHERE link_type_id = current_setting('linktypes.unknown_related')::uuid
+    AND NOT EXISTS(
+        SELECT * FROM work_item_links
+        WHERE
+            link_type_id = current_setting('linktypes.related')::uuid
+            AND source_id = l.source_id
+            AND target_id = l.target_id
+        );
+UPDATE work_item_links l 
+    SET link_type_id = current_setting('linktypes.parenting')::uuid 
+    WHERE link_type_id = current_setting('linktypes.unknown_parenting')::uuid
+    AND NOT EXISTS(
+        SELECT * FROM work_item_links
+        WHERE
+            link_type_id = current_setting('linktypes.parenting')::uuid
+            AND source_id = l.source_id
+            AND target_id = l.target_id
+        );
+
+-- Update link types in all revision entries
+UPDATE work_item_link_revisions 
+    SET work_item_link_type_id = current_setting('linktypes.bug_blocker')::uuid 
+    WHERE work_item_link_type_id = current_setting('linktypes.unknown_bug_blocker')::uuid;
+UPDATE work_item_link_revisions 
+    SET work_item_link_type_id = current_setting('linktypes.related')::uuid 
+    WHERE work_item_link_type_id = current_setting('linktypes.unknown_related')::uuid;
+UPDATE work_item_link_revisions 
+    SET work_item_link_type_id = current_setting('linktypes.parenting')::uuid 
+    WHERE work_item_link_type_id = current_setting('linktypes.unknown_parenting')::uuid;
+
+-- Update revisions
+UPDATE work_item_link_revisions rev SET
+    work_item_link_type_id = (SELECT link_type_id FROM work_item_links WHERE id = rev.work_item_link_id);
+
+-- Delete all revisions that point to a link that use the old link types 
+DELETE FROM work_item_link_revisions 
+    WHERE work_item_link_id IN (SELECT id FROM work_item_links WHERE id NOT IN (
+        current_setting('linktypes.bug_blocker')::uuid,
+        current_setting('linktypes.related')::uuid,
+        current_setting('linktypes.parenting')::uuid
+    ));
+
+-- Finally, delete links using old link types
+DELETE FROM work_item_links WHERE link_type_id NOT IN (
```

Comment:
This should not do anything at this point.. remove it just in case?

Reply:
Consider two links between the same work items but with a different link type:

```
A ----- parent of (old) -----> B
A ----- parent of (new) -----> B
``` 

I think we discussed, that we're not updating the first link but instead delete it as well as its revision, remember? The only reason I can see why we want to skip the deletion here is that the link type deletion below will cascade to all existing links with the old types.